src/main/java/com/uber/kafkaSpraynozzle/KafkaPoster.java

```diff
@@ -16,61 +23,102 @@
+    private boolean roundRobin = false;
+    private int batchSize = 20;
+    private int batchMaxBytes = 128*1024;
+    private boolean debuggingOutput = false;
 
     public KafkaPoster(
+            ConcurrentLinkedQueue<ByteArrayEntity> queue,
+            PoolingHttpClientConnectionManager cm,
+            List<String> urls,
+            ConcurrentLinkedQueue<String> logQueue,
+            KafkaFilter messageFilter
+            ){
+        this(queue, cm, urls, logQueue, messageFilter, 20, false, false );
+    }
+    public KafkaPoster(
         ConcurrentLinkedQueue<ByteArrayEntity> queue,
         PoolingHttpClientConnectionManager cm,
         List<String> urls,
         ConcurrentLinkedQueue<String> logQueue,
-        KafkaFilter messageFilter) {
+        KafkaFilter messageFilter,
+        int batchSize,
+        boolean roundRobinPost,
+        boolean debuggingOutput) {
         this.queue = queue;
         this.cm = cm;
         this.urls = urls;
         this.logQueue = logQueue;
         this.messageFilter = messageFilter;
+        this.batchSize = batchSize;
+        this.roundRobin = roundRobinPost;
+        this.debuggingOutput = debuggingOutput;
+    }
+
+    private int leastResponseIdx(int[] responseTime) {
+        int leastIdx = 0;
+        for (int i = 1; i < responseTime.length; ++i){
+            if (responseTime[i] < responseTime[leastIdx]){
+                leastIdx = i;
+            }
+        }
+        return leastIdx;
     }
 
+    private int getBatchDataSize(List<ByteArrayEntity> batch){
+        int totalSize = 0;
+        for (int i = 0; i < batch.size(); ++i){
+            totalSize += (int)batch.get(i).getContentLength();
+        }
+        return totalSize;
+    }
+
+    private CloseableHttpClient client = null;
+    private long threadId = 0L;
+    private long lastReconnect = 0L;
+    private int[] responseTime = null;
+    private long[] responseTimestamp = null;
     public void run() {
-        long threadId = Thread.currentThread().getId();
+        threadId = Thread.currentThread().getId();
         System.out.println("Starting poster thread " + threadId);
-        CloseableHttpClient client = HttpClientBuilder.create().setConnectionManager(cm).build();
-        long lastReconnect = new Date().getTime();
+        client = HttpClientBuilder.create().setConnectionManager(cm).build();
+        lastReconnect = new Date().getTime();
+        responseTime = new int[urls.size()];
+        responseTimestamp = new long[urls.size()];
+
+        List<ByteArrayEntity> batch = new ArrayList<ByteArrayEntity>(batchSize + 1);
+
+        long totalPostingTimeMs = 0L;
+        long totalPostingCount = 0L;
+        long lastReportingTime = new Date().getTime();
         while(true) {
             ByteArrayEntity jsonEntity = queue.poll();
             if(jsonEntity != null) {
                 jsonEntity = messageFilter.filter(jsonEntity);
                 if (jsonEntity != null) {
-                    try {
```

Comment:
These part are refactored into a function. Some variable are promoted to class member to reduce parameter.